Librerie/libpdel-0.6.1/pdel/structs/type/structs_type_data.3
.\" @COPYRIGHT@
.\"
.\" Author: Archie Cobbs <archie@freebsd.org>
.\"
.\" $Id: structs_type_data.3 901 2004-06-02 17:24:39Z archie $
.\"
.Dd April 22, 2002
.Dt STRUCTS_TYPE_DATA 3
.Os
.Sh NAME
.Nm STRUCTS_DATA_TYPE ,
.Nm STRUCTS_FIXEDDATA_TYPE
.Nd structs types for binary data
.Sh LIBRARY
PDEL Library (libpdel, \-lpdel)
.Sh SYNOPSIS
.In sys/types.h
.In pdel/structs/structs.h
.In pdel/structs/type/data.h
.Fn STRUCTS_DATA_TYPE charset mtype
.Fn STRUCTS_FIXEDDATA_TYPE length
.Vt extern const struct structs_type structs_type_data ;
.Vt extern const struct structs_type structs_type_hexdata ;
.Sh DESCRIPTION
The
.Fn STRUCTS_DATA_TYPE
and
.Fn STRUCTS_FIXEDDATA_TYPE
macros define a
.Xr structs 3
type (i.e., a
.Dv "struct structs_type" )
for describing variable and fixed length binary data, respectively.
.Pp
For
.Fn STRUCTS_DATA_TYPE ,
the described data structure must be a
.Dv "struct structs_data" :
.Pp
.Bd -literal -offset 3n
    struct structs_data {
	u_int   length;       /* number of bytes */
	u_char  *data;        /* bytes */
    };
.Ed
.Pp
The
.Fa length
field contains the number of bytes in the sequence.
The data itself is pointed to by
.Fa data .
.Fa mtype
is the
.Xr typed_mem 3
type used to dynamically allocate the buffer pointed to by
.Fa data .
The ASCII form of the data is a base-64 encoding using the character
set specified by
.Fa charset ,
or the default charset if
.Fa charset
is
.Dv NULL .
See
.Xr base64 3
for the definition of a valid character set.
As a special case, if
.Fa charset
is the empty string, then a straight hex encoding will be used.
This results in a longer but more human readable encoding.
.Pp
.Fn STRUCTS_FIXEDDATA_TYPE
defines a structs type for a binary sequence having fixed length
.Fa length .
The described data structure is an array of
.Fa length
.Li "u_char" .
The ASCII form is a straight hexadecimal encoding.
.Pp
Two variable length types are predefined.
.Nm structs_type_data
uses the default RFC 2045 character set and
.Nm structs_type_hexdata
uses hexadecimal encoding for the ASCII expression.
Both of these types use the
.Xr typed_mem 3
type
.Dv STRUCTS_DATA_MEMTYPE ,
defined in the header file.
.Sh SEE ALSO
.Xr base64 3 ,
.Xr libpdel 3 ,
.Xr structs 3 ,
.Xr structs_type 3 ,
.Xr typed_mem 3
.Sh HISTORY
The PDEL library was developed at Packet Design, LLC.
.Dv "http://www.packetdesign.com/"
.Sh AUTHORS
.An Archie Cobbs Aq archie@freebsd.org


Librerie/libpdel-0.6.1/pdel/structs/type/structs_type_data.c

/*
 * @COPYRIGHT@
 *
 * Author: Archie Cobbs <archie@freebsd.org>
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <net/ethernet.h>

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <limits.h>
#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <time.h>

#include "structs/structs.h"
#include "structs/type/array.h"
#include "structs/type/data.h"
#include "io/filter.h"
#include "io/base64.h"
#include "util/typed_mem.h"

#define HEXVAL(c)	(isdigit(c) ? (c) - '0' : tolower(c) - 'a' + 10)

static const char hexchars[] = "0123456789abcdef";

/*********************************************************************
			BINARY DATA TYPE
*********************************************************************/

int
structs_data_copy(const struct structs_type *type, const void *from, void *to)
{
	const char *const mtype = type->args[1].v;
	const struct structs_data *const fdata = from;
	struct structs_data *const tdata = to;
	u_char *copy;

	if (fdata->length == 0) {
		memset(tdata, 0, sizeof(*tdata));
		return (0);
	}
	if ((copy = MALLOC(mtype, fdata->length)) == NULL)
		return (-1);
	memcpy(copy, fdata->data, fdata->length);
	tdata->data = copy;
	tdata->length = fdata->length;
	return (0);
}

int
structs_data_equal(const struct structs_type *type,
	const void *v1, const void *v2)
{
	const struct structs_data *const d1 = v1;
	const struct structs_data *const d2 = v2;

	return (d1->length == d2->length
	    && memcmp(d1->data, d2->data, d1->length) == 0);
}

char *
structs_data_ascify(const struct structs_type *type,
	const char *mtype, const void *data)
{
	const char *const charmap = type->args[0].v;
	const struct structs_data *const d = data;
	struct filter *encoder;
	char *edata;
	int elen;
	int i;

	/* Handle hex encoding */
	if (charmap != NULL && *charmap == '\0') {
		if ((edata = MALLOC(mtype, d->length * 2 + 1)) == NULL)
			return (NULL);
		for (i = 0; i < d->length; i++) {
			edata[i * 2] = hexchars[(d->data[i] >> 4) & 0x0f];
			edata[i * 2 + 1] = hexchars[d->data[i] & 0x0f];
		}
		edata[i * 2] = '\0';
		return (edata);
	}

	/* Get encoding filter */
	if ((encoder = b64_encoder_create(charmap)) == NULL)
		return (NULL);

	/* Encode data */
	elen = filter_process(encoder, d->data,
	    d->length, 1, (u_char **)&edata, mtype);
	filter_destroy(&encoder);
	if (elen == -1)
		return (NULL);

	/* Return encoded data as a string */
	edata[elen] = '\0';
	return (edata);
}

int
structs_data_binify(const struct structs_type *type,
	const char *ascii, void *data, char *ebuf, size_t emax)
{
	const char *const charmap = type->args[0].v;
	const char *const mtype = type->args[1].v;
	struct structs_data *const d = data;
	struct filter *decoder;
	u_char *bdata;
	int blen;
	int i;

	/* Handle hex encoding */
	if (charmap != NULL && *charmap == '\0') {
		if ((bdata = MALLOC(mtype, (strlen(ascii) + 1) / 2)) == NULL)
			return (-1);
		for (blen = 0; *ascii != '\0'; blen++) {
			while (isspace(*ascii))
				ascii++;
			if (*ascii == '\0')
				break;
			bdata[blen] = 0;
			for (i = 4; i >= 0; i -= 4) {
				if (!isxdigit(*ascii)) {
					strlcpy(ebuf, *ascii == '\0' ?
					    "odd length hex sequence" :
					    "non-hex character seen", emax);
					FREE(mtype, bdata);
					return (-1);
				}
				bdata[blen] |= HEXVAL(*ascii) << i;
				ascii++;
			}
		}
		goto done;
	}

	/* Get decoding filter */
	if ((decoder = b64_decoder_create(charmap, 1)) == NULL)
		return (-1);

	/* Decode data */
	blen = filter_process(decoder, ascii, strlen(ascii), 1, &bdata, mtype);
	filter_destroy(&decoder);
	if (blen == -1) {
		strlcpy(ebuf, "invalid encoded binary data", emax);
		return (-1);
	}

done:
	/* Fill in structure */
	d->length = blen;
	d->data = bdata;
	return (0);
}

void
structs_data_free(const struct structs_type *type, void *data)
{
	const char *const mtype = type->args[1].v;
	struct structs_data *const d = data;

	FREE(mtype, d->data);
	memset(d, 0, sizeof(*d));
}

int
structs_data_encode(const struct structs_type *type, const char *mtype,
	struct structs_data *code, const void *data)
{
	const struct structs_data *const d = data;
	u_int32_t elength;

	if ((code->data = MALLOC(mtype, 4 + d->length)) == NULL)
		return (-1);
	elength = htonl(d->length);
	memcpy(code->data, &elength, 4);
	memcpy(code->data + 4, d->data, d->length);
	code->length = 4 + d->length;
	return (0);
}

int
structs_data_decode(const struct structs_type *type, const u_char *code,
	size_t cmax, void *data, char *ebuf, size_t emax)
{
	const char *const mtype = type->args[1].v;
	struct structs_data *const d = data;
	u_int32_t elength;

	if (cmax < 4)
		goto bogus;
	memcpy(&elength, code, 4);
	d->length = ntohl(elength);
	if (cmax < d->length) {
bogus:		strlcpy(ebuf, "encoded data is corrupted", emax);
		errno = EINVAL;
		return (-1);
	}
	if ((d->data = MALLOC(mtype, d->length)) == NULL)
		return (-1);
	memcpy(d->data, code + 4, d->length);
	return (4 + d->length);
}

const struct structs_type structs_type_data =
	STRUCTS_DATA_TYPE(NULL, STRUCTS_DATA_MEMTYPE);

const struct structs_type structs_type_hexdata =
	STRUCTS_DATA_TYPE("", STRUCTS_DATA_MEMTYPE);

/*********************************************************************
		    FIXED LENGTH BINARY DATA TYPE
*********************************************************************/

char *
structs_fixeddata_ascify(const struct structs_type *type,
	const char *mtype, const void *data)
{
	const u_char *bytes = data;
	char *s;
	int i;

	if ((s = MALLOC(mtype, type->size * 2 + 1)) == NULL)
		return (NULL);
	for (i = 0; i < type->size; i++) {
		s[i * 2] = hexchars[(bytes[i] >> 4) & 0x0f];
		s[i * 2 + 1] = hexchars[bytes[i] & 0x0f];
	}
	s[i * 2] = '\0';
	return (s);
}

int
structs_fixeddata_binify(const struct structs_type *type,
	const char *ascii, void *data, char *ebuf, size_t emax)
{
	u_char *bytes = data;
	int i;
	int j;

	for (i = 0; i < type->size; i++) {
		bytes[i] = 0;
		for (j = 4; j >= 0; j -= 4) {
			while (isspace(*ascii))
				ascii++;
			if (!isxdigit(*ascii)) {
				strlcpy(ebuf, *ascii == '\0' ?
				    "hex string is too short" :
				    "non-hex character seen", emax);
				goto fail;
			}
			bytes[i] |= HEXVAL(*ascii) << j;
			ascii++;
		}
	}
	while (isspace(*ascii))
		ascii++;
	if (*ascii != '\0') {
		strlcpy(ebuf, "hex string is too long", emax);
fail:		errno = EINVAL;
		return (-1);
	}
	return (0);
}

int
structs_fixeddata_encode(const struct structs_type *type, const char *mtype,
	struct structs_data *code, const void *data)
{
	if ((code->data = MALLOC(mtype, type->size)) == NULL)
		return (-1);
	memcpy(code->data, data, type->size);
	code->length = type->size;
	return (0);
}

int
structs_fixeddata_decode(const struct structs_type *type, const u_char *code,
	size_t cmax, void *data, char *ebuf, size_t emax)
{
	if (cmax < type->size) {
		strlcpy(ebuf, "encoded data is truncated", emax);
		errno = EINVAL;
		return (-1);
	}
	memcpy(data, code, type->size);
	return (type->size);
}

